src/app/src/main/kotlin/org/vaccineimpact/api/app/logic/CoverageLogic.kt

```diff
@@ -82,13 +120,65 @@ class RepositoriesCoverageLogic(private val modellingGroupRepository: ModellingG
                 )
             }.asSequence()
         }.toList()
+
+        val setsWithMissingRows = expectedRowLookup.missingRows()
+        if (setsWithMissingRows.any())
+        {
+            throw MissingRowsError(rowErrorMessage(setsWithMissingRows))
+        }
         touchstoneRepository.saveCoverageForTouchstone(touchstoneVersionId, records)
     }
 
-    constructor(repositories: Repositories) : this(repositories.modellingGroup,
-            repositories.responsibilities,
-            repositories.touchstone,
-            repositories.scenario)
+    private fun validate(vaccine: String,
+                         row: CoverageIngestionRow,
+                         countries: List<String>,
+                         expectedRowLookup: CoverageRowLookup): CoverageRowLookup
+    {
+        if (!expectedGAVICoverageYears.contains(row.year))
+        {
+            throw BadRequest("Unexpected year: ${row.year}")
+        }
+        if (row.activityType == ActivityType.ROUTINE)
+        {
+            val countryLookup = expectedRowLookup[vaccine] ?: generateCountryLookup(countries)
```

Comment:
So we ingest coverage for any vaccine(s) which already exist in the db. Is the expectation that GAVI will provide coverage for different vaccines with separate uploads or all in one go? If the latter, I suppose we should also be validating that there are no missing vaccines.

Reply:
Nope, explicitly need to allow for a subset of vaccines to be uploaded at one time as data is sometimes not available at the same time. The page in the portal will provide feedback on which vaccines we're still waiting on though.